src/utils.h

```diff
@@ -136,61 +136,100 @@ class ConcurrentSharedPtr {
       std::make_shared<std::recursive_mutex>();
 };
 
-template <typename I = int, typename R = unsigned char>
-struct UnionFind {
-  Vec<I> parents;
-  // we do union by rank
-  // note that we shift rank by 1, rank 0 means it is not connected to anything
-  // else
-  Vec<R> ranks;
-
-  UnionFind(I numNodes) : parents(numNodes), ranks(numNodes, 0) {
-    sequence(parents.begin(), parents.end());
+// from https://github.com/wjakob/dset, changed to add connected component
```

Comment:
Since dset uses zlib license, should that be mentioned here?  Or at least somewhere?

Reply:
In particular - "3. This notice may not be removed or altered from any source distribution."

Reply:
Agreed, and thanks for pointing that out.

Reply:
Ah, I somehow forgot restriction 3, I think I kept 1 and 2 in mind. Will fix.